rfd/XXXX-tsh-upgrade.md

```diff
@@ -0,0 +1,117 @@
+---
+authors: Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+# RFD XXXX - tsh upgrade
+
+## Required Approvers
+
+* Engineering: ...
+* Infra: ...
+* Kube: ...
+
+## What
+
+A tsh subcommand that will prompt the user to upgrade if the version of tsh is older than the cluster version.
+
+## Why
+
+One of our goals for Teleport Cloud is to increase the cadence at which clusters can be upgraded.
+This requires that Teleport Agents and Teleport client tools be up to date with the latest compatible version of the cluster.
+
+Manually upgrading Teleport Agents and client tools can be a tedious task, so customers can end up delaying their upgrades.
+Simplifying this process will help customers upgrade and stay up to date with their Teleport version.
+
+Development for Automatic Upgrades of Teleport Agents has already been completed, so the success criteria for this RFD only includes developing a simpler way to keep Teleport client tools up to date.
+
+## Details
+
+The main idea here is to add an upgrade command to the Teleport client tools. This will detect the version of the Teleport cluster.
+If the version of tsh is older than the version of the cluster, the user will be prompted to upgrade.
+The command should then re-download and re-install itself.
+
+### Identifying cluster version
+
+Detecting the version of the Teleport cluster should be straight forward.
+Teleport supports a ping RPC that returns a response containing information about the Teleport version of the cluster and the minimum client version required.
+In the case that the cluster version of Teleport is newer than the tsh version, the user will be prompted to upgrade.
+
+### Identifying installation method
```

Comment:
There's one more installation method for tsh – Teleport Connect comes bundled with it.

On Linux, the bundled version is automatically symlinked to `/usr/local/bin/tsh` if there's no existing symlink there. On Windows, the folder with the bundled tsh is automatically added to Path. On macOS you can select a menu option that adds tsh to path. See [RFD 99](https://github.com/gravitational/teleport/blob/master/rfd/0099-link-bundled-tsh.md) for more details.

If someone runs `tsh upgrade` using the binary bundled with Connect, I don't think we should upgrade the binary as it'll simply get overwritten on the next Connect update (which users have to do manually for now). But even more important, since Connect was first released we've always assumed that the version of Connect is bound to the version of tsh. As such, we didn't really put effort into making the gRPC service served by tsh daemon backwards compatible. So if someone updates the bundled tsh without updating Connect, they risk breaking the app.

I think for now when someone runs `tsh upgrade` we could detect if tsh is bundled with Connect. If so, instead of upgrading the binary we could show a message telling the user to download a new Connect version instead.

The bundled tsh lives in a specific folder in the installation folder of Connect, so for now we could see if simple pattern matching on the binary path would be sufficient.